modules/ml/src/main/java/org/apache/ignite/ml/inference/builder/IgniteDistributedInfModelBuilder.java

```diff
@@ -0,0 +1,356 @@
+        private final IgniteQueue<O> resQueue;
+
+        /** Futures that represents requests that have been sent, but haven't been responded yet. */
+        private final BlockingQueue<CompletableFuture<O>> futures = new ArrayBlockingQueue<>(QUEUE_CAPACITY);
+
+        /** Thread pool for receiver to work in. */
+        private final ExecutorService receiverThreadPool = Executors.newSingleThreadExecutor();
+
+        /** Flag identified that model is up and running. */
+        private final AtomicBoolean running = new AtomicBoolean(false);
+
+        /** Receiver future. */
+        private volatile Future<?> receiverFut;
+
+        /**
+         * Constructs a new instance of distributed inference model.
+         *
+         * @param ignite Ignite instance.
+         * @param suffix Suffix that with correspondent templates formats service and queue names.
+         * @param reader Inference model reader.
+         * @param parser Inference model parser.
+         * @param instances Number of service instances maintaining to make distributed inference.
+         */
+        DistributedInfModel(Ignite ignite, String suffix, InfModelReader reader, InfModelParser<I, O> parser,
+            int instances) {
+            this.ignite = ignite;
+            this.suffix = suffix;
+
+            reqQueue = ignite.queue(String.format(INFERENCE_REQ_QUEUE_NAME_PATTERN, suffix), QUEUE_CAPACITY, queueCfg);
+            resQueue = ignite.queue(String.format(INFERENCE_RES_QUEUE_NAME_PATTERN, suffix), QUEUE_CAPACITY, queueCfg);
+
+            startReceiver();
+            startService(reader, parser, instances);
+
+            running.set(true);
+        }
+
+        /** {@inheritDoc} */
+        @Override public Future<O> predict(I input) {
+            if (!running.get())
+                throw new IllegalStateException("Inference model is not running");
+
+            reqQueue.put(input);
+
+            try {
+                CompletableFuture<O> fut = new CompletableFuture<>();
+                futures.put(fut);
+                return fut;
+            }
+            catch (InterruptedException e) {
+                close(); // In case of exception in the above code the model state becomes invalid and model is closed.
+                throw new RuntimeException(e);
+            }
+        }
+
+        /**
+         * Starts Apache Ignite services that represent distributed inference infrastructure.
+         *
+         * @param reader Inference model reader.
+         * @param parser Inference model parser.
+         * @param instances Number of service instances maintaining to make distributed inference.
+         */
+        private void startService(InfModelReader reader, InfModelParser<I, O> parser, int instances) {
+            ignite.services().deployMultiple(
+                String.format(INFERENCE_SERVICE_NAME_PATTERN, suffix),
+                new IgniteDistributedInfModelService<>(reader, parser, suffix),
+                instances,
```

Comment:
are you sure that totalCnt == maxPerNodeCnt ?

Reply:
Good point. I moved this parameter on API level, please have a look.